### ISSUE TITLE
Elementum restarting issue

### ISSUE THREAD

**Issue by RunAway189** [StefanIlchev/ElementumService#12]
Hello,

First of all thanks for your amazing work. I had everything working perfectly but recently I had to restore my device. I've installed everything back as usual but now the "Elementum daemon has started" pops up when I launch the movie, then after few seconds the movie stops playing. It seems the daemon crashes.

I'm on NVIDIA Shield 2019 Tube firmware 9.0.2 (33.1.0.319) Android 11; Kodi 19.4

**Comment by StefanIlchev** [comment]
Are you using the latest version? (0.1.860040 as of writing)

Could you share the log from after this happens by executing the following command:

```bat
adb logcat -d service.elementum.android
```

**Comment by RunAway189** [comment]
> Are you using the latest version? (0.1.860040 as of writing)
> 
> Could you share the log from after this happens by executing the following command:
> 
> ```batchfile
> adb logcat -d service.elementum.android
> ```

Yes, I'm on the latest version. So just to be sure, do O have to give the logcat command after the issue happens, right? Once confirmed I'll post it here, thanks for your help

**Comment by StefanIlchev** [comment]
Yes after, so we can see what was logged when it happened.

**Comment by RunAway189** [comment]
Perfect, here it is. Let me know if you need anything else
[logcat.txt](https://github.com/StefanIlchev/ElementumService/files/8456598/logcat.txt)

**Comment by RunAway189** [comment]
Weird, I don't have any of these settings enabled. It seems it's a shield firmware bug then, no other reason. I'll try again with the next update. Thanks anyway for your help :)

**Comment by ppongoo** [comment]
Hi. I am happy to find your work and appreciate it.
I am running your elementum(0.1.860040) on both android 11 and 10. On android 11, it works perfectly fine. But android 10, it has a different restarting issue. When I first launch elementum service, it works ok. But on the next launch, it doesn't start elementum daemon.  I am using an android 10 device which my IPTV provider gave me. That device's OS doesn't support notifications. I disabled powersaving modes to solve this problem but it still doesn't help. So what I am doing is I manually kill the elementum service process before I start Kodi everytime.

So I am wondering if I can add a code that as the kodi started it kills the elementum service first and launchs it again?
I wish to attach log files but any adb commands didn't work.

**Comment by ninormalen** [comment]
> Hi. I am happy to find your work and appreciate it. I am running your elementum(0.1.860040) on both android 11 and 10. On android 11, it works perfectly fine. But android 10, it has a different restarting issue than OP's. When I first launch elementum service, it works ok. But on the next launch, it doesn't start elementum daemon. I am using an android 10 device which my IPTV provider gave me. That device's OS doesn't support notifications. I disabled powersaving modes to solve this problem but it still didn't help. So what I am doing is I manually kill the elementum service process before I start Kodi everytime.
> 
> So I am wondering if I can add a code that as the kodi started it kills the elementum service first and launchs it again? I wish to attach log files but any adb commands didn't work.

Hi, same problem on Chromecast Google tv. I solved problem as delete data ad cash on elementum service every time on start service. I appreciate to solve the problem.

**Comment by StefanIlchev** [comment]
@ppongoo, @ninormalen I've already modified the addon to also signal the service app when the "still starting up" message appears so simply interacting with it should be enough to get it running. Why do you feel the need to stop the service right before starting it again? Although pointless leaving it running between usage sessions should be fine.

**Comment by ppongoo** [comment]
@StefanIlchev  When I first start the addon, I get a message that "the daemon has started", but the message doesn't appear on the second run. It's kinda weird because even though I couldn't get that message, I still can start the addon and get magnets from torrent providers, but sadly it fails to download torrents. What's more strange is that if I close Kodi without downloading any torrent, the next run will work fine.  To be clear, once I downloaded any torrent, I should shut down the Elementum service in the background in order to work at the next run.

**Comment by StefanIlchev** [comment]
The "daemon has started" message won't appear if the service is already running when you open Kodi which it probably is seeing as you are able to use the addon. I can't think of a reason why the addon wouldn't work in this case. I tried leaving the service running in the background while restarting Kodi by force-stopping it from settings and it seemed to be working fine. The only way for the service to keep running after Kodi stops that I can think of is if Kodi wasn't gracefully shut down from it's UI and was instead force-stopped from settings, or dismissed from recent apps, or killed by the system for resources, etc.

**Comment by ninormalen** [comment]
> Sorry for my late reponse. I was out of town for a while. But I accidently found what caused the issue. I set an extra trackers to 'None' and it finally functions normally. I usually shut down Kodi through UI so I pretty much bet that tracker setting was the problem.

Can tell me where is that option. Thank you.

**Comment by ppongoo** [comment]
> > Sorry for my late reponse. I was out of town for a while. But I accidently found what caused the issue. I set an extra trackers to 'None' and it finally functions normally. I usually shut down Kodi through UI so I pretty much bet that tracker setting was the problem.
> 
> Can tell me where is that option. Thank you.

[Elementum - Settings - BitTorent - Add extra Trackers]
I changed it to "None".

**Comment by RunAway189** [comment]
@StefanIlchev Hey man, I just wanted to report that I tried to reinstall the v0.1.860027 and now the crashing issue is finally fixed, no more restart after few seconds. I think it may be related to the feature "the addon will now try to start the service again if it's not responding." introduced in 029 but I'm just guessing, I'm not an expert. Thanks anyway for the help :)

**Comment by StefanIlchev** [comment]
@RunAway189 that doesn't make sense unless starting the daemon is what stops playback. I tried stopping and starting the service during playback and removing the code starting the service again if it's not responding and the result was the same - if the daemon is stopped during playback then playback will stop once it runs out of buffer regardless of whether the daemon is started again.

**Comment by hhromic** [comment]
> > > Sorry for my late reponse. I was out of town for a while. But I accidently found what caused the issue. I set an extra trackers to 'None' and it finally functions normally. I usually shut down Kodi through UI so I pretty much bet that tracker setting was the problem.
> > 
> > 
> > Can tell me where is that option. Thank you.
> 
> [Elementum - Settings - BitTorent - Add extra Trackers] I changed it to "None".

Nvidia Shield 2019 Pro user here, had the exact symptoms you described in your first post (Torrents not working the second time Kodi is launched) and changing "Add extra Trackers" to "None" also solved it for me. Weird.

**Comment by lomax** [comment]
> Hello,
> 
> First of all thanks for your amazing work. I had everything working perfectly but recently I had to restore my device. I've installed everything back as usual but now the "Elementum daemon has started" pops up when I launch the movie, then after few seconds the movie stops playing. It seems the daemon crashes.
> 
> I'm on NVIDIA Shield 2019 Tube firmware 9.0.2 (33.1.0.319) Android 11; Kodi 19.4
> 
> EDIT: it seems it happens only with 4K HDR content

I've managed to find a workaround for this issue with HDR content play leading to elementum service breaking. I've decreased number of connection for bittorent even futher from 20 set before, to 10 which seems ok for my wired shield connection. Also I've set a limit on number of connections per second to 5 instead of auto and the issue appeared on the first minute of playing HDR content is gone. Haven't checked for the full film yet.

**Comment by RunAway189** [comment]
> > Hello,
> > First of all thanks for your amazing work. I had everything working perfectly but recently I had to restore my device. I've installed everything back as usual but now the "Elementum daemon has started" pops up when I launch the movie, then after few seconds the movie stops playing. It seems the daemon crashes.
> > I'm on NVIDIA Shield 2019 Tube firmware 9.0.2 (33.1.0.319) Android 11; Kodi 19.4
> > EDIT: it seems it happens only with 4K HDR content
> 
> I've managed to find a workaround for this issue with HDR content play leading to elementum service breaking. I've decreased number of connection for bittorent even futher from 20 set before, to 10 which seems ok for my wired shield connection. Also I've set a limit on number of connections per second to 5 instead of auto and the issue appeared on the first minute of playing HDR content is gone. Haven't checked for the full film yet.

Thanks a lot. Sadly the issue is back on my device so it wasn't depending on the service version, @StefanIlchev was right. I've tried your suggestion as well without any success. At this point I have to blame my shield tube for its low amount of RAM, there is no other explanation. This is the only issue I'm having with it and it's so disappointing. The only "solution" is to use xTorrent and streaming the content to JustPlayer while downloading the file on my SD card

**Comment by cudaspriest** [comment]
Hi, just to report that I have the same problem on the full Shield (with more RAM).

**Comment by RunAway189** [comment]
> Hi, just to report that I have the same problem on the full Shield (with more RAM).

Thanks, I'm glad it doesn't depend on that either

**Comment by Makaone** [comment]
Hi,

Thanks for the great work. I'm using Latest Nvidia shield with your latest release. Everything works great but the daemon starts and stops continuously.

**Comment by StefanIlchev** [comment]
@Makaone could you share the log from after this happens?

**Comment by Makaone** [comment]
@StefanIlchev are you referring to the kodi log file?

**Comment by StefanIlchev** [comment]
Actually I was referring to Android's log which you can get with the following ADB command:

```bat
adb logcat -d
```

**Comment by Makaone** [comment]
Hi, I have tried to use adb shell to obtain the log but I keep getting errorKind regardsOn 3 Oct 2022, at 10:46 am, StefanIlchev ***@***.***> wrote:﻿
Actually I was referring to Android's log which you can get with the following ADB command:
adb logcat -d

**Comment by Makaone** [comment]
mdarcy:/ $ adb logcat -d
/system/bin/sh: inaccessible or not found

**Comment by Makaone** [comment]
@StefanIlchev thanks for clarifying. I can get the log now, but I'm just trying to work out how to save it to file from the shield. I don't have a pc or know how to connect via adb even if I borrowed one. But I continue to have service restarts all the time. Any help would be awesome. I have disable the plugin so it doesn't cause me problems.

**Comment by StefanIlchev** [comment]
You can redirect the output to a file on the device like so:

```bat
logcat -d >/sdcard/Download/logcat.log
```

Then you'll need to access it somehow. You can install File Commander from the Play Store. It can share files/folders in your local network. Just open the URL it gives you in an internet browser on any device.

**Comment by Makaone** [comment]
@StefanIlchev here is a link to the log file

https://www.dropbox.com/s/nc0snnjoqppcgw0/logcat.log?dl=0

**Comment by StefanIlchev** [comment]
@Makaone I see the service starting once and no indication of it stopping. Is it the plugin that restarts?

Could you please try exiting Kodi and then force stopping both it and the service app by executing the following commands:

```bat
am force-stop org.xbmc.kodi
```

```bat
am force-stop service.elementum.android
```

Then clear the log by executing the following command:

```bat
logcat -c
```

After that start Kodi, wait for the daemon to start and do whatever causes it to restart.

Finally, dump the log in a file once again by executing the following command and share it:

```bat
logcat -d >/sdcard/Download/logcat.log
```

**Comment by Makaone** [comment]
@StefanIlchev thanks for your prompt reply. I completed what you suggested and below is the latest log.

https://www.dropbox.com/s/s72sn69vv7y9thy/logcat.log?dl=0

When I started kodi the elementum service started. I started a download using elementum and after a minute or so the notification said elementum still warming up, and then the pop up saying the service had started appeared again. Then it will start downloading again.

I hope this helps

**Comment by StefanIlchev** [comment]
@Makaone once again I can only see "Starting Elementum daemon" logged once. I found only one message containing the word "warming" and it gets posted to show up after 30 seconds if something hasn't finished loading.

Is this message showing up disruptive in any other way than being an eyesore? Is the plugin usable before this message shows up? Does it stop being usable after it shows up?

**Comment by Makaone** [comment]
@StefanIlchev in the last log everything had started correctly. Once I had selected a file to Watch/download it started fine, but a couple of minutes in, the screen froze and crashed back to the kodi Home Screen. Then a pop saying "elementum cache is still warming up" appeared, followed by "elementum service has started" then the file started downloading again according to the notification bar that appears on the screen showing name of movie and speed etc.
This happens even if I select a file to download but not watch it at the same time. It may crash 2 or three times during a download.
I hope this helps in some way.

**Comment by Makaone** [comment]
@StefanIlchev the last log I sent was straight after the crash happened

**Comment by StefanIlchev** [comment]
@Makaone I'm sorry but I looked through the new log you provided and the old one but I still can't find a cause for your issue. One thing that I did find odd is that the plugin never logs the daemon's start args like so:

`WARNING <general>: [plugin.video.elementum] elementumd: start args: ['N/A', `...

And it's not like it doesn't log anything as both your files contain:

`WARNING <general>: [plugin.video.elementum] Cleaning up dialogs`
`WARNING <general>: [plugin.video.elementum] Resetting RPC objects...`

For a sanity check could you:
1. stop Kodi gracefully from the power options;
2. wait for the service app to stop if it hasn't by the time you see the launcher screen;
3. launch Kodi again and wait for the "elementum service has started" notification;
4. dump logcat to a file.

You'll know you've done it correctly if you can find `SUBPROCESS_EXIT_VALUE` in the file.

**Comment by Makaone** [comment]
@StefanIlchev i have tried everything to try and and create the 
SUBPROCESS_EXIT_VALUE to no avail...
Is there way to stop the notification of the service starting as this seems to be when the crashes occur?

https://www.dropbox.com/s/qfzp9hkkz1e4lpa/logcat%20%283%29.log?dl=0

**Comment by StefanIlchev** [comment]
@Makaone sanity check failed, I guess. What you're reporting makes no sense to me.

If SUBPROCESS_EXIT_VALUE never gets logged that means the daemon process never stops and it needs to stop in order to start again which is the only time it notifies that "Elementum daemon has started" from what I can tell by looking at its code. Also, as I previously commented the plugin never logs the daemon's start.

If the device has so little resources it starts forcefully killing foreground services, it would explain the lack of SUBPROCESS_EXIT_VALUE but not the lack of "Starting Elementum daemon" in the logs which precedes "Elementum daemon has started" notifications.

I'm going to need more information to resolve this.

**Comment by StefanIlchev** [comment]
@Makaone maybe there's nothing in the log because it gets logged before the tail end that `logcat -d` dumps. You can try logging some restarts with logcat in parallel by executing the following command:

```shell
logcat -c; logcat >/sdcard/Download/logcat.log & pid=$!; read; kill $pid
```

To stop it enter:

```shell
some non-empty text
```

**Comment by Makaone** [comment]
@StefanIlchev i have done what you said and run the adb command, then started kodi, waited for everything to start including the elementum service. Selected a torrent and everything worked ok for a couple of minutes, then I received the "elementum service has started" notification about 4 times in a minute or so, then the file that was playing crashed back to the file selection page. 
Elementum kept downloading according to the notification at the bottom showing speed etc.

I hope this log helps

https://www.dropbox.com/s/y9dozoirbyb5d5s/logcat%20%284%29.log?dl=0

**Comment by Makaone** [comment]
@StefanIlchev here is another log inclusive of shut down and restarts of kodi. 

https://www.dropbox.com/s/6t8q53qr3olvc8z/logcat%20%285%29.log?dl=0

**Comment by StefanIlchev** [comment]
@Makaone thanks, these look a bit saner.

In the last log I can see that the plugin made a shutdown request but the daemon never stopped. I can't say if that's how it originally worked and it doesn't matter because the system would've killed all processes anyway. Also, I noticed that on start the plugin looks for traces of a previous daemon process to kill which it can no longer do as it belongs to a different app.

I released a new version in which the daemon will now be restarted instead of reused if running when the plugin starts as I no longer believe it is intended to work with multiple sessions. Hopefully a restart will also solve your notification issue.

By the way, when you write "elementum service has started" I assume you mean "Elementum daemon has started" because I can't find such a message.

**Comment by Makaone** [comment]
@StefanIlchev thanks heaps for your great work. 
So now the daemon will only start when a file is selected to watch/download and then stop once the file is completed. Is that right?

Will I be able to select multiple downloads at the same time etc?

Also to stop the logging, you said to put non empty text! Can you please clarify.

I will try the new daemon when I get home. Once again thanks for your perseverance.

**Comment by Makaone** [comment]
@StefanIlchev i have updated to the latest but everything happening the same. Latest log

https://www.dropbox.com/s/qkskjk0fvejyyoy/logcat%20%286%29.log?dl=0

**Comment by StefanIlchev** [comment]
> @StefanIlchev thanks heaps for your great work. So now the daemon will only start when a file is selected to watch/download and then stop once the file is completed. Is that right?

No, the daemon will start when the plugin starts even if it has to be stopped first. This is closer to how it originally worked as a subprocess of Kodi that couldn't continue after the app was killed. Before if the daemon was already running with the same arguments I did nothing.

> Will I be able to select multiple downloads at the same time etc?

Yes.

> Also to stop the logging, you said to put non empty text! Can you please clarify.

Just type `whatever` to stop logging otherwise it'll just keep going. Ending the session might also stop it which mustn't happen while trying to record.

> @StefanIlchev i have updated to the latest but everything happening the same. Latest log
> 
> https://www.dropbox.com/s/qkskjk0fvejyyoy/logcat%20%286%29.log?dl=0

I can't find `plugin.video.elementum` anywhere in this one.

> @StefanIlchev here is another log inclusive of shut down and restarts of kodi.
> 
> https://www.dropbox.com/s/6t8q53qr3olvc8z/logcat%20%285%29.log?dl=0

Could you please do what you did here with the new version. I expect to see `SUBPROCESS_EXIT_VALUE`.

**Comment by Makaone** [comment]
@Stefanllchev here is an updated log inclusive of 3 restarts.https://www.dropbox.com/s/kqgf8khs2zwaquy/logcat%20%287%29.log?dl=0Sent from iPhoneOn 18 Oct 2022, at 3:26 pm, StefanIlchev ***@***.***> wrote:﻿

@StefanIlchev thanks heaps for your great work. So now the daemon will only start when a file is selected to watch/download and then stop once the file is completed. Is that right?

No, the daemon will start when the plugin starts even if it has to be stopped first. This is closer to how it originally worked as a subprocess of Kodi that couldn't continue after the app was killed. Before if the daemon was already running with the same arguments I did nothing.

Will I be able to select multiple downloads at the same time etc?

Yes.

Also to stop the logging, you said to put non empty text! Can you please clarify.

Just type whatever to stop logging otherwise it'll just keep going. Ending the session might also stop it which mustn't happen while trying to record.

@StefanIlchev i have updated to the latest but everything happening the same. Latest log
https://www.dropbox.com/s/qkskjk0fvejyyoy/logcat%20%286%29.log?dl=0

I can't find plugin.video.elementum anywhere in this one.

@StefanIlchev here is another log inclusive of shut down and restarts of kodi.
https://www.dropbox.com/s/6t8q53qr3olvc8z/logcat%20%285%29.log?dl=0

Could you please do what you did here with the new version. I expect to see SUBPROCESS_EXIT_VALUE.

**Comment by StefanIlchev** [comment]
@Makaone thanks, I can see them in the log and they look as expected. Now on to the main problem, can you do the following in this order:
1. start logging;
2. start Kodi;
3. cause the plugin/daemon to crash and restart at least once;
4. stop Kodi;
5. start Kodi;
6. stop Kodi;
7. stop logging;
8. upload the log.

**Comment by Makaone** [comment]
@StefanIlchev i have followed your steps and here is the log.

"Elementum daemon has started appeared twice while watching the file before it crashed back to kodi main screen

https://www.dropbox.com/s/wn90206l5ih8o6n/logcat%20%288%29.log?dl=0

**Comment by StefanIlchev** [comment]
@Makaone the log is incomplete. My guess is that it ends somewhere between steps 2 and 3. When you stopped logging was there a message that implied that a process wasn't found/killed/stopped?

I've modified the parallel logging command to redirect the output from attempting to kill logcat to the same file:

```shell
logcatLog=/sdcard/Download/logcat.log; logcat -c; logcat >$logcatLog & pid=$!; read; echo '\n--------- beginning of kill logcat' >>$logcatLog; kill $pid >>$logcatLog 2>&1; echo '\n---------    ending of kill logcat' >>$logcatLog
```

Now somewhere around the end of the file there should be two lines containing `beginning of kill logcat` and `ending of kill logcat` respectively. There also shouldn't be a line like `/system/bin/sh: kill: 123456789: No such process` anywhere between them. If there is then the logging wasn't successful.

Could you please try again with the modified command.

**Comment by Makaone** [comment]
@StefanIlchev i completed exactly as you said. Started kodi, selected file. It started downloading/ playing, crashed back to kodi Home Screen and kept downloading at the bottom.
Stopped kodi, started kodi and waited for "Elementum daemon has started" closed kodi.

Here is the log. I hope it helps.

https://www.dropbox.com/s/vk0b1emgw3r1dzk/logcat%20%289%29.log?dl=0

**Comment by StefanIlchev** [comment]
@Makaone unfortunately this one cuts off some time before you stop Kodi the first time, just like the last one. If you look at the size of the last two, they are a lot smaller compared to the one before.

**Comment by Makaone** [comment]
@StefanIlchev do you having any ideas as to why it's not logging the whole process? I've been doing everything as requested.

**Comment by StefanIlchev** [comment]
@Makaone I don't know. I've had it running for hours writing hundreds of megabytes of text without a problem.

Have you tried as a test to just start logging and stop it after a second to see if the `kill logcat` lines show up?

**Comment by mixspano** [comment]
> Hello,
> 
> First of all thanks for your amazing work. I had everything working perfectly but recently I had to restore my device. I've installed everything back as usual but now the "Elementum daemon has started" pops up when I launch the movie, then after few seconds the movie stops playing. It seems the daemon crashes.
> 
> I'm on NVIDIA Shield 2019 Tube firmware 9.0.2 (33.1.0.319) Android 11; Kodi 19.4
> 
> EDIT: it seems it happens only with 4K HDR content

Hi, 
I had the same problem with my nvidia shield pro 2019.
Best solution I found was to go to settings --> services --> smb client --> client --> minimum protocol version --> SMBv2
                                                                                                                                 --> maximum protocol version --> SMBv3
That resolved the problem.
My system is Nvidia Shield Pro 2019 , firmware 9.1.0 (33.2.0.125), Android 11, Kodi 19.4

**Comment by dnicotra** [comment]
Same issue here, with an Nvidia shield pro. Any news?

**Comment by StefanIlchev** [comment]
> Same issue here, with an Nvidia shield pro. Any news?

Not really, I still need logs to do something if it even is an issue with the modifications I made.

**Comment by dnicotra** [comment]
> > Same issue here, with an Nvidia shield pro. Any news?
> 
> Not really, I still need logs to do something if it even is an issue with the modifications I made.

Unfortunately, I am not close to the shield right now. But when I'm back home after the Christmas break I can try the same procedure.

**Comment by StefanIlchev** [comment]
I worked around the issue with [this](https://github.com/StefanIlchev/ElementumService/commit/fb01b459add9d46b4913b4221ebc91a10a2d3719#diff-03e554d12a75d5603d4a2c4bda5c30e35dbfce04590265fcab7a023a1ddad48aL261) change. Now even if there is a restart it should not take so long as to stop playback on most devices.
